\documentclass[11pt,oneside,letterpaper]{article}

% graphicx package, useful for including eps and pdf graphics
\usepackage{graphicx}
\DeclareGraphicsExtensions{.png,.png,.jpg}

% basic packages
\usepackage{color}
\usepackage{parskip}
\usepackage{float}
\usepackage{microtype}
\usepackage{url}
\usepackage{hyperref}

% text layout
\usepackage{geometry}
\geometry{textwidth=17cm} % 15.25cm for single-space, 16.25cm for double-space
\geometry{textheight=22.5cm} % 22cm for single-space, 22.5cm for double-space

% helps to keep figures from being orphaned on a page by themselves
\renewcommand{\topfraction}{0.85}
\renewcommand{\textfraction}{0.1}

% bold the 'Figure #' in the caption and separate it with a period
% Captions will be left justified
\usepackage[labelfont=bf,labelsep=period,font=small]{caption}

% cite package, to clean up citations in the main text. Do not remove.
\usepackage{cite}

\usepackage{authblk}
\renewcommand\Authands{ \& }
\renewcommand\Authfont{\normalsize \bf}
\renewcommand\Affilfont{\small \normalfont}
\makeatletter
\renewcommand\AB@affilsepx{, \protect\Affilfont}
\makeatother

\title{Integrative forecasting of seasonal influenza A/H3N2 by genotype and phenotype}
%\title{Experimentally informed forecasts of seasonal influenza A/H3N2}
%\title{Long-term forecasts of seasonal influenza reveal historical contingency of fitness metrics}

\author[1,2]{John Huddleston}
\author[2]{Richard A.\ Neher}
\author[1]{Trevor Bedford}

\affil[1]{Vaccine and Infectious Disease Division, Fred Hutchinson Cancer Research Center, Seattle, WA, USA}
\affil[2]{Moleculary and Cell Biology, University of Washington, Seattle, WA, USA}
\affil[3]{Biozentrum, University of Basel, Basel, Switzerland}

\begin{document}

\maketitle

\section{Abstract}

\section{Introduction}

Seasonal influenza virus infects 5--15\% of the global population every year causing an estimated 250,000 to 500,000 deaths annually \cite{flufactsheet}.
Vaccination remains the most effective public health response available.
However, frequent viral mutation results in viruses that escape previously acquired human immunity.
The World Health Organization (WHO) selects vaccine viruses to match circulating viruses, but because the process of vaccine development and distribution requires several months to complete, accurate vaccine strain selection requires a prediction of which viruses will predominate approximately one year after vaccine viruses are selected.
Current vaccine predictions favor viruses that are distinct from prior vaccine viruses in the hemagglutinin (HA) protein, which acts as the primary target of human immunity.
The hemagglutination inhibition (HI) assay \cite{hirst1943studies} is used to measure the degree of cross-reactivity between pairs of circulating viruses.
HI assays are fundamental for vaccine strain selection, but they are laborious and low-throughput compared to genome sequencing \cite{Wood:2012ii}.
As a result, researchers have developed computational methods to predict influenza fitness from sequence data alone \cite{Luksza:2014hj,Steinbruck:2014kq,Neher:2014eu}.

Despite the promise of these sequence-only models, they explicitly omit experimental measurements of antigenic or functional phenotypes.
Recent developments in computational methods and influenza virology have made it feasible to integrate these important metrics of influenza fitness into a single predictive model.
For example, phenotypic measurements of antigenic drift are now accessible through phylogenetic models \cite{Neher:2016hy} and functional phenotypes for HA are available from deep mutational scanning experiments \cite{Lee2018}.
We describe an approach to integrate previously disparate sequence-only models of influenza evolution with high-quality experimental measurements of antigenic drift and functional constraint.

\section{Results}

\subsection{Recapitulation of Luksza and Lassig 2014}

We implemented a predictive model inspired by the basic framework of {\L}uksza and L\"assig \cite{Luksza:2014hj}.

\subsection{Forecasts with experimental and phylogenetic fitness metrics}

\subsection{Historical contingency of fitness metrics}

\subsection{Forecasts of other influenza genes and subtypes by A/H3N2 hemagglutinin models}

\subsection{Comparison of chosen and optimal vaccine strains}

\subsection{Forecasts for February 2020}

\section{Discussion}

\section{Methods}

\subsection{Strain selection}

\subsection{Multiple sequence alignment}

\subsection{Phylogenetic interference}

\subsection{Frequency estimation}

\subsection{Fitness model}

\subsection{Model fitting and evaluation}

\subsubsection{Model targets}

\subsubsection{Time-series cross-validation}

\subsection{Fitness metrics}

\subsubsection{Antigenic drift}

\subsubsection{Functional constraint}

\subsubsection{Clade growth}

\clearpage

\bibliographystyle{nih}
\bibliography{manuscript}

\end{document}
