\documentclass[stdletter,letterpaper,addrfromright,orderfromdateto,dateleft,11pt,noaddrto,sigleft]{newlfm}
\topmarginskip{-0.0in}
\bottommarginskip{-0.0in}
\leftmarginsize{1.25in}
\rightmarginsize{1.5in}
\sigskipbefore{0.2in}
\sigskipafter{0in}
\noLines
\nolines
\noHeadline
\noheadline
\signature{John Huddleston}

\namefrom{}
\addrfrom{
	\includegraphics[width=6.5cm]{figures/fhcrc_logo} \\
  Vaccines and Infectious Disease Division \\
  Fred Hutchinson Cancer Research Center \\
  1100 Fairview Ave N \\
  Seattle, WA 98109, USA}
\emailfrom{jlhudd@uw.edu}

\greetto{Dear Editors,}
\closeline{\\Sincerely,}

% comments
\usepackage{color}
\usepackage{ulem}
\definecolor{purple}{rgb}{0.459,0.109,0.538}
\def\tb#1#2{\sout{#1} \textcolor{purple}{#2}}
\def\tbc#1{\textcolor{purple}{[#1]}}

\begin{document}

\begin{newlfm}
  Please find attached our manuscript entitled ``Integrating genotypes and phenotypes improves long-term forecasts of seasonal influenza A/H3N2 evolution''.
  We would be grateful if you considered it for publication in \textit{eLife}.

  The World Health Organization (WHO) selects vaccine strains for seasonal influenza A/H3N2 based on experimental measurements of antigenic drift and computational models of hemagglutinin (HA) sequence evolution.
  Hemagglutination inhibition (HI) assays provide phenotypic measurements of antigenic drift that have been the gold standard for decades.
  The WHO augments these data with computational fitness models based on HA sequences including a mechanistic model by {\L}uksza and L{\"a}ssig (2014) that predicts the success of H3N2 populations one year into the future and a phylogenetic model by Neher et al. (2014) called the local branching index (LBI) that estimates H3N2 strain fitness based on rapid branching in HA phylogenies.
  No published studies to date have integrated all of these genetic and phenotypic measures of HA into a single model to understand their relative contributions to influenza fitness.

  To address this gap, we developed a fitness framework that integrates genetic and phenotypic measures of antigenic drift and functional constraint with phylogenetic measures of clade growth.
  We implemented novel measures of antigenic drift from HI assays, functional constraint from deep mutational scanning (DMS) experiments, and clade growth from recent changes in clade frequencies.
  We also reimplemented {\L}uksza and L{\"a}ssig's sequence-based measures of antigenic drift and functional constraint and Neher et al.'s LBI metric.
  We rigorously tested the performance of these fitness metrics individually and in relevant combinations using time-series cross-validation.
  The most accurate models included a combination of sequence-based functional constraint and either HI-based measures of antigenic drift or LBI-based measures of clade growth.

  Our findings address several open questions in the field of long-term influenza forecasting and point the way for future forecasting efforts.
  We find that genetic estimates of antigenic drift are insufficient replacements for the gold standard phenotypic measures provided by HI assays.
  We further show that the success of previous models based on these genetic estimates is due, in part, to inadvertently borrowing information from the future.
  In contrast, simple genetic estimates of functional constraint are superior to more detailed estimates provided by deep mutational scanning (DMS) measurements from a single background strain.
  These results suggest that future forecasting models should identify metrics that are the least likely to be affected by historical contingency.

  We designed our fitness framework as an open source, reproducible workflow built on tidy data frames and integrated it into Nextstrain, a real-time pathogen surveillance tool.
  As our models estimate the sequence composition of future populations instead of phylogenetic clade frequencies, they could enable forecasting efforts for other rapidly evolving organisms whose genomes are not suited for phylogenetic inference such as recombining viruses and bacteria.
  This manuscript is relevant to an audience of evolutionary biologists, infectious disease researchers, and computational biologists.
  For this reason, we would like to submit this manuscript for consideration in the Special Issue on Evolutionary Medicine.
\end{newlfm}

\end{document}
